### PR TITLE
feat(obstacle_cruise_planner): support pointcloud-based obstacles

### PR DESCRIPTION
**Pull request by mitukou1109**
## Description

Make `obstacle_cruise_planner` compatible with `sensor_msgs::msg::PointCloud2`-based obstacles.

The node subscribes to a `sensor_msgs::msg::PointCloud2` topic and converts the point cloud into obstacles by the following steps:
1. transform the cloud into the trajectory frame
2. downsample the cloud by voxel grid and apply Euclidean clustering to get candidates of obstacles
3. check if each point in a cluster is within the stop / slow down margin
4. when any points are inside the stop / slow down area, create a new obstacle object containing:
    - the nearest point from the ego for stop obstacles
    - the nearest and the farthest from the ego for slow down obstacles

Stop and slow down planning are performed with the same trajectory generation algorithm as for ML-based (`autoware_perception_msgs::msg::PredictedObject`) obstacles.

Cruising and yielding are not supported since point cloud-based obstacles have no velocity information.

Margin adjustment on curves (https://github.com/autowarefoundation/autoware.universe/pull/4952) is also unsupported due to the following reason:
- Point cloud-based obstacles do not have shape information.
- Unlike PredictedObject polygons, point clouds do not necessarily capture the entire obstacle. This could lead the margin to be smaller than in the ML-based case.

launch PR: https://github.com/autowarefoundation/autoware_launch/pull/980

## Tests performed

Psim

```
ros2 launch autoware_launch planning_simulator.launch.xml map_path:=$HOME/autoware_map/sample-map-planning vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit
```

with the following changes to obstacle_cruise_planner.param.yaml:

```
/**:
  ros__parameters:
    common:
     stop_obstacle_type:
        unknown: false
        car: false
        truck: false
        bus: false
        trailer: false
        motorcycle: false
        bicycle: false
        pedestrian: false
        pointcloud: true

     slow_down_obstacle_type:
        unknown: false
        car: false
        truck: false
        bus: false
        trailer: false
        motorcycle: false
        bicycle: false
        pedestrian: false
        pointcloud: true

    behavior_determination:
      pointcloud_search_radius: 5.0
      pointcloud_voxel_grid_x: 0.05
      pointcloud_voxel_grid_y: 0.05
      pointcloud_voxel_grid_z: 100000.0
      pointcloud_cluster_tolerance: 1.0
      pointcloud_min_cluster_size: 1
      pointcloud_max_cluster_size: 100000
```

[Slow down / stop planning test](https://github.com/autowarefoundation/autoware.universe/assets/50359861/801c6577-d32a-4eac-9e8a-a4fc1e17faeb)

Evaluator:
- w/ point cloud handling disabled: https://evaluation.tier4.jp/evaluation/reports/a6893361-4d9d-567f-8a2b-e92f49148286?project_id=prd_jt
- w/ both ML-based objects and point cloud handling enabled: https://evaluation.tier4.jp/evaluation/reports/1a144b5e-943e-53d0-b7ab-7ea016e29f48?project_id=prd_jt

## Effects on system behavior

It is confirmed that the changes do not affect the behavior when the point cloud handling is disabled.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/obstacle_cruise_planner/include/obstacle_cruise_planner/common_structs.hpp [hunk 1]
```diff
@@ -87,6 +117,8 @@ struct Obstacle
   std::vector<PredictedPath> predicted_paths;
   double ego_to_obstacle_distance;
   double lat_dist_from_obstacle_to_traj;
+  PointCloud pointcloud;
+  bool pointcloud_repr;
```
Comment (satoshi-ota) [inline planning/obstacle_cruise_planner/include/obstacle_cruise_planner/common_structs.hpp:121]: Could you explain the purpose of this variable?

File: planning/obstacle_cruise_planner/include/obstacle_cruise_planner/common_structs.hpp [hunk 2]
```diff
@@ -66,15 +70,41 @@ struct Obstacle
...
   }
 
-  Polygon2d toPolygon() const { return tier4_autoware_utils::toPolygon2d(pose, shape); }
+  Obstacle(const rclcpp::Time & arg_stamp, const PointCloud & object)
+  : stamp(arg_stamp), uuid(""), pointcloud(object), pointcloud_repr(true)
+  {
+    twist.linear.x = twist.linear.y = twist.angular.z = 0.0;
+
+    Eigen::Vector4d centroid;
+    pcl::compute3DCentroid(pointcloud, centroid);
```
Comment (satoshi-ota) [inline planning/obstacle_cruise_planner/include/obstacle_cruise_planner/common_structs.hpp:88]: Why did you calculate centroid of pointcloud?

File: planning/autoware_obstacle_cruise_planner/include/autoware/obstacle_cruise_planner/common_structs.hpp [hunk 3]
```diff
@@ -58,25 +58,42 @@ struct Obstacle
     const geometry_msgs::msg::Pose & arg_pose, const double ego_to_obstacle_distance,
     const double lat_dist_from_obstacle_to_traj)
   : stamp(arg_stamp),
+    ego_to_obstacle_distance(ego_to_obstacle_distance),
+    lat_dist_from_obstacle_to_traj(lat_dist_from_obstacle_to_traj),
```
Comment (satoshi-ota) [inline planning/autoware_obstacle_cruise_planner/include/autoware/obstacle_cruise_planner/common_structs.hpp:62]: Can you remove these diff...? I'd like to reduce differential as much as possible.

File: planning/autoware_obstacle_cruise_planner/include/autoware/obstacle_cruise_planner/common_structs.hpp [hunk 4]
```diff
@@ -58,25 +58,42 @@ struct Obstacle
...
+    stop_collision_point(stop_collision_point),
+    slow_down_front_collision_point(slow_down_front_collision_point),
+    slow_down_back_collision_point(slow_down_back_collision_point)
+  {
+  }
 
   rclcpp::Time stamp;  // This is not the current stamp, but when the object was observed.
+  double ego_to_obstacle_distance;
+  double lat_dist_from_obstacle_to_traj;
```
Comment (satoshi-ota) [inline planning/autoware_obstacle_cruise_planner/include/autoware/obstacle_cruise_planner/common_structs.hpp:94]: same as https://github.com/autowarefoundation/autoware.universe/pull/6907/files#r1646746224

File: planning/autoware_obstacle_cruise_planner/config/obstacle_cruise_planner.param.yaml [hunk 5]
```diff
@@ -68,8 +69,10 @@
         motorcycle: true
         bicycle: true
         pedestrian: true
+        pointcloud: false
 
     behavior_determination:
+      use_pointcloud: false
```
Comment (satoshi-ota) [inline planning/autoware_obstacle_cruise_planner/config/obstacle_cruise_planner.param.yaml:75]: I think we don't need this param.

File: planning/autoware_obstacle_cruise_planner/src/node.cpp [hunk 6]
```diff
@@ -387,12 +432,17 @@ ObstacleCruisePlannerNode::ObstacleCruisePlannerNode(const rclcpp::NodeOptions &
...
+  tf_buffer_ = std::make_unique<tf2_ros::Buffer>(get_clock());
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
   const auto longitudinal_info = LongitudinalInfo(*this);
 
   ego_nearest_param_ = EgoNearestParam(*this);
 
   enable_debug_info_ = declare_parameter<bool>("common.enable_debug_info");
   enable_calculation_time_info_ = declare_parameter<bool>("common.enable_calculation_time_info");
+  use_pointcloud_ = declare_parameter<bool>("common.use_pointcloud");
```
Comment (satoshi-ota) [inline planning/autoware_obstacle_cruise_planner/src/node.cpp:445]: Can you remove this variable `use_pointcloud_`...? I guess the logical product of use_pointcloud_for_stop_ and use_pointcloud_for_slow_down_ represents same meaning.

File: planning/autoware_obstacle_cruise_planner/src/node.cpp [hunk 7]
```diff
@@ -871,9 +1109,14 @@ std::optional<CruiseObstacle> ObstacleCruisePlannerNode::createCruiseObstacle(
...
   const Obstacle & obstacle, const double precise_lat_dist)
 {
-  const auto & object_id = obstacle.uuid.substr(0, 4);
   const auto & p = behavior_determination_param_;
 
+  if (use_pointcloud_) {
+    return std::nullopt;
+  }
+
+  const auto & object_id = obstacle.uuid.substr(0, 4);
```
Comment (satoshi-ota) [inline planning/autoware_obstacle_cruise_planner/src/node.cpp:1118]: Please reduce diff if possible.
```suggestion
  if (use_pointcloud_) {
    return std::nullopt;
  }
  
  const auto & object_id = obstacle.uuid.substr(0, 4);
  const auto & p = behavior_determination_param_;
```

File: planning/autoware_obstacle_cruise_planner/src/node.cpp [hunk 8]
```diff
@@ -914,10 +1157,15 @@ std::optional<CruiseObstacle> ObstacleCruisePlannerNode::createCruiseObstacle(
 std::optional<CruiseObstacle> ObstacleCruisePlannerNode::createYieldCruiseObstacle(
   const Obstacle & obstacle, const std::vector<TrajectoryPoint> & traj_points)
 {
+  const auto & p = behavior_determination_param_;
```
Comment (satoshi-ota) [inline planning/autoware_obstacle_cruise_planner/src/node.cpp:1160]: same as above.

**Review by satoshi-ota — COMMENTED** [review]
@mitukou1109 Good work :+1: 
But could you exclude diffrential of `dummy_perception_publisher`? I want to forcus on `obstacle_cruise_planner` in this PR.

File: planning/autoware_obstacle_cruise_planner/include/autoware/obstacle_cruise_planner/common_structs.hpp (hunk 3 above)
Comment (mitukou1109) [reply planning/autoware_obstacle_cruise_planner/include/autoware/obstacle_cruise_planner/common_structs.hpp:62]: This is for suppressing "will be initialized after" warning. These 2 member variables are common to both PredictedObject and PointCloud, so I thought it would be better to move them above the other members for PredictedObject.

Comment (satoshi-ota) [reply planning/autoware_obstacle_cruise_planner/include/autoware/obstacle_cruise_planner/common_structs.hpp:62]: OK. Thanks.

File: planning/autoware_obstacle_cruise_planner/src/node.cpp (hunk 7 above)
Comment (mitukou1109) [reply planning/autoware_obstacle_cruise_planner/src/node.cpp:1117]: https://github.com/autowarefoundation/autoware.universe/pull/6907/commits/0c913b239d1ee9f54a2374c6e0b2c5a42fac78cf

File: planning/autoware_obstacle_cruise_planner/src/node.cpp (hunk 6 above)
Comment (mitukou1109) [reply planning/autoware_obstacle_cruise_planner/src/node.cpp]: Fixed: https://github.com/autowarefoundation/autoware.universe/pull/6907/commits/f2dc8cf5dc8e8521abf7009124b33171a52a9533

**Comment by mitukou1109** [comment]
Removed changes of dummy_perception_publisher: https://github.com/autowarefoundation/autoware.universe/pull/6907/commits/64a3191c82fe3cee6b167b0001f5b892c6a68110, https://github.com/autowarefoundation/autoware.universe/pull/6907/commits/d5260f751db42978a6eb193a066371857c59c27d, https://github.com/autowarefoundation/autoware.universe/pull/6907/commits/d8c8dbb9d4e905c45ed2577f8e422a3bdd72b7e3

File: planning/autoware_obstacle_cruise_planner/src/node.cpp [hunk 9]
```diff
@@ -495,13 +549,15 @@ void ObstacleCruisePlannerNode::onTrajectory(const Trajectory::ConstSharedPtr ms
 {
   const auto ego_odom_ptr = ego_odom_sub_.takeData();
   const auto objects_ptr = objects_sub_.takeData();
+  const auto pointcloud_ptr = pointcloud_sub_.takeData();
```
Comment (yuki-takagi-66) [inline planning/autoware_obstacle_cruise_planner/src/node.cpp:552]: Could you change to call this function only if necessary? Computing cost of this function is not small enough

**Review by yuki-takagi-66 — CHANGES_REQUESTED** [review]
Thank you for your works.
May I ask you to achieve handling both of predicted objects and point cloud?
I feel current implementation handle both with exclusively.

File: planning/autoware_obstacle_cruise_planner/src/node.cpp [hunk 10]
```diff
@@ -520,11 +582,12 @@ void ObstacleCruisePlannerNode::onTrajectory(const Trajectory::ConstSharedPtr ms
   //    (1) with a proper label
   //    (2) in front of ego
   //    (3) not too far from trajectory
-  const auto target_obstacles = convertToObstacles(ego_odom, objects, traj_points);
+  const auto target_obstacles =
+    convertToObstacles(ego_odom, objects_ptr, pointcloud_ptr, traj_points, msg->header);
```
Comment (yuki-takagi-66) [inline planning/autoware_obstacle_cruise_planner/src/node.cpp:586]: Wouldn't it be cleaner to call convertToObstacles()  by type? like
const auto target_obstacles =
convertToObstacles(ego_odom, objects_ptr, traj_points, msg->header) + convertToObstacles(ego_odom, pointcloud_ptr, traj_points, msg->header);

I think my misunderstanding comment may have gotten you to make this change. Sorry.

File: planning/autoware_obstacle_cruise_planner/src/node.cpp [hunk 11]
```diff
@@ -871,9 +1052,14 @@ std::optional<CruiseObstacle> ObstacleCruisePlannerNode::createCruiseObstacle(
...
   const Obstacle & obstacle, const double precise_lat_dist)
 {
-  const auto & object_id = obstacle.uuid.substr(0, 4);
+  if (obstacle.source_type == Obstacle::SourceType::POINTCLOUD) {
+    return std::nullopt;
+  }
+
   const auto & p = behavior_determination_param_;
 
+  const auto & object_id = obstacle.uuid.substr(0, 4);
```
Comment (yuki-takagi-66) [inline planning/autoware_obstacle_cruise_planner/src/node.cpp:1061]: unnecessary change

File: planning/autoware_obstacle_cruise_planner/src/node.cpp [hunk 12]
```diff
@@ -914,10 +1100,15 @@ std::optional<CruiseObstacle> ObstacleCruisePlannerNode::createCruiseObstacle(
 std::optional<CruiseObstacle> ObstacleCruisePlannerNode::createYieldCruiseObstacle(
   const Obstacle & obstacle, const std::vector<TrajectoryPoint> & traj_points)
 {
+  if (obstacle.source_type == Obstacle::SourceType::POINTCLOUD) {
+    return std::nullopt;
+  }
+
+  const auto & p = behavior_determination_param_;
```
Comment (yuki-takagi-66) [inline planning/autoware_obstacle_cruise_planner/src/node.cpp:1107]: unnecessary change

File: planning/autoware_obstacle_cruise_planner/src/node.cpp (hunk 10 above)
Comment (mitukou1109) [reply planning/autoware_obstacle_cruise_planner/src/node.cpp]: Fixed: https://github.com/autowarefoundation/autoware.universe/pull/6907/commits/5d5ec3f6d0b213d2ee5501b202870a35c2807beb

File: planning/autoware_obstacle_cruise_planner/src/node.cpp [hunk 13]
```diff
@@ -1182,7 +1388,10 @@ std::optional<StopObstacle> ObstacleCruisePlannerNode::createStopObstacle(
   const auto & object_id = obstacle.uuid.substr(0, 4);
 
   // NOTE: consider all target obstacles when driving backward
-  if (!isStopObstacle(obstacle.classification.label)) {
+  if (
+    (obstacle.source_type == Obstacle::SourceType::PREDICTED_OBJECT &&
+     !isStopObstacle(obstacle.classification.label)) ||
+    (obstacle.source_type == Obstacle::SourceType::POINTCLOUD && !use_pointcloud_for_stop_)) {
```
Comment (satoshi-ota) [inline planning/autoware_obstacle_cruise_planner/src/node.cpp:1394]: [imo] it's a little bit difficult to understand this condition.

This is my suggestion.

```c++
  if (
    (obstacle.source_type == Obstacle::SourceType::PREDICTED_OBJECT &&
     !isStopObstacle(obstacle.classification.label)) {
    return std::nullopt;
  }

  if (
    (obstacle.source_type == Obstacle::SourceType::POINTCLOUD && !use_pointcloud_for_stop_)) {
    return std::nullopt;
  }
```

File: planning/autoware_obstacle_cruise_planner/src/node.cpp [hunk 14]
```diff
@@ -1270,122 +1496,150 @@ std::optional<SlowDownObstacle> ObstacleCruisePlannerNode::createSlowDownObstacl
...
+  if (obstacle.source_type == Obstacle::SourceType::PREDICTED_OBJECT) {
+    slow_down_condition_counter_.addCurrentUuid(obstacle.uuid);
+  }
 
-  if (!enable_slow_down_planning_ || !isSlowDownObstacle(obstacle.classification.label)) {
+  if (
+    !enable_slow_down_planning_ ||
+    (obstacle.source_type == Obstacle::SourceType::PREDICTED_OBJECT &&
+     !isSlowDownObstacle(obstacle.classification.label)) ||
+    (obstacle.source_type == Obstacle::SourceType::POINTCLOUD && !use_pointcloud_for_slow_down_)) {
```
Comment (satoshi-ota) [inline planning/autoware_obstacle_cruise_planner/src/node.cpp:1508]: same as https://github.com/autowarefoundation/autoware.universe/pull/6907/files#r1656411386.

**Review by yuki-takagi-66 — APPROVED** [review]

**Review by satoshi-ota — APPROVED** [review]
